docs(*): added proposal of app metrics (#4185)
kumahq/kuma: docs/proposals/app-metrics.md

# Application metrics
 
## Context
 
Currently, when an application exposes metrics as regular service it's protected and traffic is secured (mTLS enabled), it's a bit tricky

> to [expose those metrics without mTLS](https://kuma.io/docs/dev/policies/traffic-metrics/#expose-metrics-from-applications). Also, by configuring paths user's can by mistake expose more than it's necessary. To avoid it we should come up with a solution that fulfills all the requirements: scraping metrics with/without mTLS and safer exposing of metrics.

Reviewer: 1. You need to have Prometheus per Mesh
2. Envoy needs to build a listener for every INSTANCE in order for a transparent proxy to work which produces a gigantic configuration

 
## Requirements
* support different application ports/paths
* support more than one container/application
* support no mTLS communication with Prometheus
 
## Idea
 

> The idea is to create an aggregator in `kuma-dp`, to be specific Metrics Hijacker, that will be responsible for the aggregation of metrics from applications defined to scrap(within local env: pod/vm). Communication with `kuma-dp` might be configured as secure but it won't be required(and not recommended). It's worth mentioning that this idea can support only Prometheus endpoint scrapping and exposing metrics as Prometheus.

Reviewer 2: ```suggestion
Instead of exposing application metrics directly, we propose creating a metrics aggregator in `kuma-dp`. The aggregator will be responsible for the scraping of metrics from applications proxied by this instance of `kuma-dp`.
`kuma-dp` can then expose this port with or without mTLS. It's worth mentioning that this idea can support only Prometheus endpoint scraping and exposing metrics as Prometheus.
```

Coming from someone who knows very little about the idea, this I think is clearer about what we're doing (if I understand correctly).

 
### Universal
At the mesh level, it will be possible to define the default exposed path and port at which metrics are going to be accessible.
```
type: Mesh
name: default
metrics:
 enabledBackend: prometheus-1
 backends:
 - name: prometheus-1
   type: prometheus
   conf:

>     skipMTLS: false

Reviewer: for smooth migration, we probably need to support both for 2 versions. Not sure if we should bother if the end goal is to have a separate policy

    port: 5670 # default exposed port with metrics 
    path: /metrics
    tags:   
      kuma.io/service: dataplane-metrics
```
`Dataplane` configuration enables at which port/path `kuma-dp` is going to expose aggregated metrics from Envoy and other applications running within one workflow. Also, it's possible to override the `Mesh` configuration of endpoints to scrap by the `Dataplane` configuration. They are going to be identified by `name` and if `Mesh` and `Dataplane` configurations have definitions with the same name, then the `Dataplane` configuration has precedence before `Mesh` and is merged.
```
type: Dataplane
mesh: default
name: redis-1
networking:
 address: 192.168.0.1
 inbound:

>  - port: 9000

Reviewer 2: Can we make the example a little more realistic? When I see `redis` I expect an inbound for `6379` and when I don't I stumble and then I'm not sure whether that's because of what's in the proposal, etc.

Reviewer 3: Maybe using just redis as a name is not right

   servicePort: 1234
   tags:
     kuma.io/service: backend
 metrics:
   type: prometheus
   conf:
     skipMTLS: true
     port: 1234
     path: /non-standard-path

>      aggregate:

Reviewer 2: I think we should highlight in the above text that this is where you define which metrics the aggregator scrapes. If this is shared with the `Mesh` `conf`, does having this make sense on the `Mesh` level?

Author: Yes, it makes sense to have it on mesh level and override/append on when someone adds annotation. In most of the cases, it might be 90% of apps configured by mesh resource, and few can override.

         kuma.io/service: dataplane-metrics
```
In the case of Kubernetes configuration of endpoints to scrap is done by label `prometheus.metrics.kuma.io/aggregate-{name}-port` and `prometheus.metrics.kuma.io/aggregate-{name}-path`, where `{name}` is used for distinguish between config for different containers/apps running in the pod.
 
```
apiVersion: apps/v1
kind: Deployment
metadata:
 name: example-app
 namespace: kuma-example
spec:
 ...
 template:
   metadata:
     ...
     labels:
       # indicate to Kuma that this Pod doesn't need a sidecar
       kuma.io/sidecar-injection: enabled
     annotations:
       prometheus.metrics.kuma.io/aggregate-app-port: 1234 # format is aggregate-{name}-port
       prometheus.metrics.kuma.io/aggregate-app-path: "/metrics"
       prometheus.metrics.kuma.io/aggregate-opa-port: 12345 # format is aggregate-{name}-port
       prometheus.metrics.kuma.io/aggregate-opa-path: "/stats"
   spec:
     containers:
       ...
```
 
## Use existing prometheus tags

Prometheus on K8s uses annotation `prometheus.io/{path/port/scrapping}`, we might use it to map them for kuma specifc 

> annotations that are going to be used to scrap endpoints. At the begging we might not implement this feature but in future we might get this.

Reviewer 3: ```suggestion
annotations that are going to be used to scrap endpoints. To start we will not implement this feature but in future we might get this.
```

metrics:
 enabledBackend: prometheus-1
 backends:
 - name: prometheus-1
   type: prometheus
   conf:
     skipMTLS: true
     mtls:
       require: true
       tags: # tags that can be referred in Traffic Permission when metrics are secured by mTLS 
         kuma.io/service: dataplane-metrics
     aggregate:
     - name: app
       port: 1234
       path: /metrics # defaults to /metrics
     - name: opa
       port: 8888
       path: /metrics # defaults to /metrics
```
`Dataplane` configuration enables at which port/path `kuma-dp` is going to expose aggregated metrics from Envoy and other applications running within one workflow. Also, it's possible to override the `Mesh` configuration of endpoints to scrap by the `Dataplane` configuration. They are going to be identified by `name` and if `Mesh` and `Dataplane` configurations have definitions with the same name, then the `Dataplane` configuration has precedence before `Mesh` and is merged.
```
type: Dataplane
mesh: default
name: redis-1
networking:
 address: 192.168.0.1
 inbound:
 - port: 9000
   servicePort: 1234
   tags:
     kuma.io/service: backend
 metrics:
   type: prometheus
   conf:
     skipMTLS: true
     port: 1234

>      path: /non-standard-path

Reviewer 3: Should this be a DP option?

Author: I think we could allow it on `Mesh` and allow to override on DP mode - for Universal.

Reviewer 3: Do we still do something at the Mesh level?

Reviewer: IMHO having aggregate on the mesh level makes sense. If 90% of apps have a standard port and path to expose path, probably it's useful to define this on the Mesh level and solve the remaining 10% on the DP level

Reviewer 3: Ok!

>      path: /non-standard-path

Reviewer 3: Should this be a DP option?

Author: I think for now it's going to be easier to add data to dataplane object. Creating a new policy would be probably better but we also working on new policies so it might be blocked for a while.

     aggregate:
     - name: app # overrides app from Mesh
       port: 1236
       path: /metrics # defaults to /metrics
     # inherits opa from Mesh
```
 
### Kubernetes
 
The same `Mesh` configuration works for K8s and Universal.
```
apiVersion: kuma.io/v1alpha1
kind: Mesh
metadata:
 name: default
spec:
 metrics:
   enabledBackend: prometheus-1
   backends:
   - name: prometheus-1
     type: prometheus
     conf:
       skipMTLS: false
       port: 5670
       path: /metrics
       tags: # tags that can be referred in Traffic Permission when metrics are secured by mTLS 
         kuma.io/service: dataplane-metrics
       aggregate:
       - name: app
         port: 1236
         path: /metrics # defaults to /metrics
       - name: opa
         port: 8888

>          path: /metrics # defaults to /metrics

Reviewer 3: To me exposing this in the mesh atm is not really a usable solution (your apps within a mesh are just too different).

I think we should keep it as dataplane options and introduce a policy later for this.

Reviewer 3: Ok then :)

>          path: /metrics # defaults to /metrics

Reviewer 3: Only concern is this work being stuck on the new policy work landing.

```
In the case of Kubernetes configuration of endpoints to scrap is done by label `prometheus.metrics.kuma.io/aggregate-{name}-port` and `prometheus.metrics.kuma.io/aggregate-{name}-path`, where `{name}` is used for merging config with one defined in `Mesh`.
 
```
apiVersion: apps/v1
kind: Deployment
metadata:
 name: example-app
 namespace: kuma-example
spec:
 ...
 template:
   metadata:
     ...
     labels:
       # indicate to Kuma that this Pod doesn't need a sidecar

>        kuma.io/sidecar-injection: disabled

Reviewer 3: Probably need enabled here ;)

>        kuma.io/sidecar-injection: disabled

Reviewer 3: Probably need enabled here ;)
